evm_arithmetization/src/fixed_recursive_verifier.rs

```diff
@@ -868,6 +959,86 @@ where
         }
     }
 
+    /// Create two-to-one block aggregation circuit.
+    ///
+    /// # Arguments
+    ///
+    /// - `block_circuit`: circuit data for the block circuit, that constitues
+    ///   the base case for aggregation.
+    ///
+    /// # Outputs
+    ///
+    /// Returns a [`TwoToOneBlockCircuitData<F, C, D>`].
+    fn create_two_to_one_block_circuit(
+        block_circuit: &BlockCircuitData<F, C, D>,
+    ) -> TwoToOneBlockCircuitData<F, C, D>
+    where
+        F: RichField + Extendable<D>,
+        C: GenericConfig<D, F = F>,
+        C::Hasher: AlgebraicHasher<F>,
+    {
+        let mut builder = CircuitBuilder::<F, D>::new(block_circuit.circuit.common.config.clone());
+
+        let mix_hash = builder.add_virtual_hash();
+        builder.register_public_inputs(&mix_hash.elements);
+
+        // We need to pad by PIS to match the count of PIS of the `base_proof`.
+        let mut padding = block_circuit.circuit.common.num_public_inputs;
+        // The number of PIS that will be added *after* padding by
+        // [`add_verifier_data_public_inputs()`].
+        padding -= verification_key_len(&block_circuit.circuit);
+        // Account for `mix_pv_hash`.
+        padding -= builder.num_public_inputs();
+
+        for _ in 0..padding {
+            let target = builder.zero();
+            builder.register_public_input(target);
+        }
+
+        let user_pis_len = builder.num_public_inputs();
+        let cyclic_vk = builder.add_verifier_data_public_inputs();
+
+        let lhs = Self::add_agg_child(&mut builder, &block_circuit.circuit);
+        let rhs = Self::add_agg_child(&mut builder, &block_circuit.circuit);
+
+        let lhs_public_values = lhs.public_values_vec(&mut builder);
+        let rhs_public_values = rhs.public_values_vec(&mut builder);
+
+        let lhs_agg_pv_hash = extract_two_to_one_block_hash(&lhs_public_values);
+        let rhs_agg_pv_hash = extract_two_to_one_block_hash(&rhs_public_values);
+
+        let lhs_base_pv_hash = builder
+            .hash_n_to_hash_no_pad::<C::InnerHasher>(lhs_public_values[..user_pis_len].to_owned())
+            .elements;
+        let rhs_base_pv_hash = builder
+            .hash_n_to_hash_no_pad::<C::InnerHasher>(rhs_public_values[..user_pis_len].to_owned())
+            .elements;
```

Comment:
`[..user_pis_len]` is redundant as this corresponds to the length of the vec returned by `public_values_vec()` (see related comment there).

Reply:
Additionally, why do we hash the entire set of public inputs? This is a bit wasteful. The VK is fixed, in the case we do not have an aggregated proof, we should only hash the actual public values, and nothing else.

Reply:
> to not include the verifier key in the hash, but only the blockproof's inputs.

That is my point. You're padding with 2197 zero targets, requiring to hash right after `user_pis_len` = 2201 elements. That represents 276 Poseidon permutations, while for `aggregated` proofs (i.e. proofs that already wrapped 2 block proofs together), we don't need to do so, because once aggregated, we already have a hash of the PublicValues available.
This is essentially why I also found approach 3 (my suggestion) cleaner, as it only hashes what's necessary.

Reply:
Right, let me get this right: it's redundant, not in the sense that I can avoid it in the current approach, but in the sense that I can use the other approach to avoid it. Is that right? I cannot see how I can resolve this issue otherwise.

Reply:
@Nashtare I think I get it now. It was wrong. I was hashing more than just the Public Values.

Please check latest commit and let's sync before resolving this.

Reply:
FWIW, the number are equal in the end:

```
[evm_arithmetization/src/fixed_recursive_verifier.rs:999:9] PublicValuesTarget::SIZE = 2201
[evm_arithmetization/src/fixed_recursive_verifier.rs:1000:9] builder.num_public_inputs() = 2201
```

Your performance concern is valid in any case.